freestyle-effects/shared/src/main/scala/effects/validation.scala

```diff
@@ -0,0 +1,63 @@
+package freestyle.effects
+
+import freestyle._
+import cats.{MonadState}
+import cats.data.{State, Validated, ValidatedNel, NonEmptyList}
+
+
+object validation {
+  /** A validation exception with an explanation. **/
+  trait ValidationException {
+    def explanation: String
+  }
+
+
+  case class NotValid(explanation: String) extends ValidationException
+
+  object ValidationException {
+    def apply(msg: String): ValidationException = NotValid(msg)
+  }
+
+  type Errors = List[ValidationException]
+
+  /** An algebra for introducing validation semantics in a program. **/
+  @free sealed trait ValidationM[F[_]] {
+    def valid[A](x: A): FreeS.Par[F, A]
+
+    def invalid(err: ValidationException): FreeS.Par[F, Unit]
+
+    def errors: FreeS.Par[F, Errors]
+
+    def fromEither[A](x: Either[ValidationException, A]): FreeS.Par[F, Either[ValidationException, A]]
+
+    def fromValidatedNel[A](x: ValidatedNel[ValidationException, A]): FreeS.Par[F, ValidatedNel[ValidationException, A]]
+  }
+
+  object implicits {
+    implicit def freeStyleValidationMStateInterpreter[M[_]](
+      implicit
+        MS: MonadState[M, Errors]
```

Comment:
@peterneyens Would this face the same issues described in https://github.com/47deg/freestyle/issues/85 ?